File: app/src/main/java/com/mouredev/weeklychallenge2022/Challenge27.dart
/*
 * Reto #27
 * VECTORES ORTOGONALES
 * Fecha publicación enunciado: 07/07/22
 * Fecha publicación resolución: 11/07/22
 * Dificultad: FÁCIL
 *
 * Enunciado: Crea un programa que determine si dos vectores son ortogonales.
 * - Los dos array deben tener la misma longitud.
 * - Cada vector se podría representar como un array. Ejemplo: [1, -2]
 *
 * Información adicional:
 * - Usa el canal de nuestro discord (https://mouredev.com/discord) "🔁reto-semanal" para preguntas, dudas o prestar ayuda a la comunidad.
 * - Puedes hacer un Fork del repo y una Pull Request al repo original para que veamos tu solución aportada.
 * - Revisaré el ejercicio en directo desde Twitch el lunes siguiente al de su publicación.
 * - Subiré una posible solución al ejercicio el lunes siguiente al de su publicación.
 *
 */
void main(List<String> args) {
  List<int> vec1 = [2, 6];
  List<int> vec2 = [-3, 1];

  bool areOrthogonal(List<int> v1, List<int> v2) {
    if (v1.length != v2.length) return false;
    int result = 0;
    for (var i = 0; i < v1.length; i++) result += v1[i] * v2[i];
    return result == 0;
  }

  bool areOrt = areOrthogonal(vec1, vec2);
  print("Los vectores $vec1 y $vec2${areOrt ? "" : " NO"} son ortogonales");
}

File: app/src/main/java/com/mouredev/weeklychallenge2022/Challenge27.kt
package com.mouredev.weeklychallenge2022

/*
 * Reto #27
 * VECTORES ORTOGONALES
 * Fecha publicación enunciado: 07/07/22
 * Fecha publicación resolución: 11/07/22
 * Dificultad: FÁCIL
 *
 * Enunciado: Crea un programa que determine si dos vectores son ortogonales.
 * - Los dos array deben tener la misma longitud.
 * - Cada vector se podría representar como un array. Ejemplo: [1, -2]
 *
 * Para determinar que son ortogonales es necesario que el producto punto (o producto escalar) de
 * ambos vectores sea igual a cero: U.V = 0. Entonces resolvemos de la siguiente manera:
 * tenemos U = (1, 2) y V = (-2, 1), debemos hallar U · V. Quedaría de esta manera:
 * U · V = (1)*(-2) + (2)*(1) = 0
 */

fun main (){
    val vecOne = arrayOf(4, -2, 1)
    val vecTwo = arrayOf(1, 2, 0)
    println(areOrthogonal(vecOne, vecTwo))
}

fun areOrthogonal(vecOne : Array<Int>, vecTwo : Array<Int>) : Boolean{

    if (vecOne.size != vecTwo.size){
        return false
    }else{
        var sum = 0
        for (i in vecOne.indices){
            sum += vecOne[i]*vecTwo[i]
        }
        return sum == 0
    }
}
